Add `revng-model-graph` tool

AlviseDeFaveri opened the pull request:
This MR introduces a `ModelGraphPrinter` class whose purpose is to generate a `.dot` representation of a given `model::Type`.

Features:
- `Type.dumpDotOnFile("filename")` to be used e.g. inside `gdb`
- `revng model graph <model or LLVM>`

Example of the output (`.dot` file, produced from `printf/x86-64/canonicalized.bc`
[model.txt](https://github.com/revng/revng/files/8826284/model.txt)

aleclearmind commented on lib/Model/ModelGraphPrinter.cpp:
```diff
@@ -0,0 +1,235 @@
+}
+
+ModelGraphPrinter::ModelGraphPrinter(llvm::raw_ostream &Out,
+                                     bool OrthoEdges /*= false*/,
+                                     bool DisplayUnionsVertically /*= false*/) :
```
```suggestion
                                     bool OrthoEdges,
                                     bool DisplayUnionsVertically) :
```

aleclearmind commented on lib/Model/ModelGraphPrinter.cpp:
```diff
@@ -0,0 +1,235 @@
+  // Label
+  Out << "[label=\"";
+
+  for (auto &Qual : QT.Qualifiers) {
+    switch (Qual.Kind) {
+    case model::QualifierKind::Array:
+      Out << "Array[" << Qual.Size << "]";
+      break;
+
+    case model::QualifierKind::Pointer:
+      Out << "Pointer (size " << Qual.Size << ")";
+      break;
+
+    default:
+      break;
+    }
+
+    Out << ",\\n";
+  }
```
There's a spurious comma at the end.
Something like:
```suggestion
  const char *Prefix = "";
  for (auto &Qual : QT.Qualifiers) {
    Out << Prefix;
    Prefix = ",\\n";
    switch (Qual.Kind) {
    case model::QualifierKind::Array:
      Out << "Array[" << Qual.Size << "]";
      break;

    case model::QualifierKind::Pointer:
      Out << "Pointer (size " << Qual.Size << ")";
      break;

    default:
      break;
    }
  }
```

aleclearmind commented on include/revng/Model/ModelGraphPrinter.h:
```diff
@@ -0,0 +1,55 @@
+
+#include "revng/Model/QualifiedType.h"
+
+/// Use this class to print a graph representing one or more model types
+/// and their subtypes.
```
I'm impressed by your discipline in documentation!

AlviseDeFaveri replied:
thanks!

aleclearmind commented:
Pretty cool!

The code prints the graph of the types, not of the model in general. Can you:
1. Rename the tool to `revng-model-export-type-graph`. We have `revng model import ...`, I'd also like `revng model export ...`.
2. `s/ModelGraphPrinter/TypeSystemPrinter/`

aleclearmind commented on tools/model/graph/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+revng_add_executable(revng-model-graph Main.cpp)
+
+llvm_map_components_to_libnames(LLVM_LIBRARIES IRReader BitWriter)
```
Why `BitWriter`?

AlviseDeFaveri replied:
Leftover from copypasta, sorry

AlviseDeFaveri commented:
Latest version looks like this
![Screenshot from 2022-06-20 19-24-53](https://user-images.githubusercontent.com/26742412/174652901-0afb26c5-2ab4-4af3-b68b-ca834d210860.png)

aleclearmind commented on include/revng/Model/Type.h:
```diff
@@ -85,6 +85,7 @@ class model::Type : public model::generated::Type {
   bool verify(bool Assert) const debug_function;
   RecursiveCoroutine<bool> verify(VerifyHelper &VH) const;
   void dump() const debug_function;
+  void dumpDotOnFile(const char *FileName) const debug_function;
```
* @pfez was saddened by the fact that the graph doesn't allow to easily get to a function prototype.
What do you think of moving this to `model::Binary` and include a node for each `model::Function` and an edge for the prototype and another for the stack frame.
We could also do this for `model::Segment`s, once antonio's stuff is in.
* Rename the method to `dumpTypeGraph`, call the argument "Path" and use `StringRef` instead of `const char *`.

pfez replied:
Small aside, I believe using `const char *` is a trick for enabling to call the function from `gdb`. If that's the case we could have 2 overloads, one with `const char *` and one with `StringRef`. Does this make sense?

aleclearmind replied:
@pfez right! Let's leave it like it is.

AlviseDeFaveri replied:
1. yes good idea thanks for the suggestion. I also realized typedefs are missing, so I'll add those.
2. fez is right

AlviseDeFaveri commented:
- Print Typedefs
- Print `model::Function`s
- Renamed `Type::dumpDotOnFile` into `Type::dumpTypeGraph`
- Added `Function::dumpTypeGraph`
- Added `Binary::dumpTypeGraph` (prints all functions and type)

Looks like this

![Screenshot from 2022-07-06 19-53-23](https://user-images.githubusercontent.com/26742412/177612929-f27d1d59-e5ab-4ac3-9510-11500b19f31e.png)
